chrome/content/zotero/xpcom/utilities_internal.js

```diff
@@ -32,6 +32,282 @@
+			}
+		}
+		
+		// If both items are translated
+		if (item1 && item2) {
+			// Combine item1 and item2
+			for (let key in item2) {
+				if (Zotero.Utilities.fieldIsValidForType(key, item1.itemType)) {
+					// Add all item2 fields that doesn't exist in item1
+					if (!item1[key] && item2[key]) {
+						item1[key] = item2[key];
+					}
+				}
+			}
+			
+			// If item1 doesn't have creators, take them from item2
+			if ((!item1.creators || !item1.creators.length) && item2.creators) {
+				item1.creators = item2.creators;
+			}
+			
+			newItem = item1;
+		}
+		// If at least one item was translated
+		else {
+			newItem = item1 || item2;
+		}
+		
+		return newItem;
+	},
+	
+	
+	/**
+	 * Combines an existing item with the new metadata.
+	 * The existing item usually has additional data like notes,
+	 * tags, creation date, etc. therefore we can't completely
+	 * replace it by creating a new item
+	 *
+	 * @param {Zotero.Item} item
+	 * @param {Object} newItem Item metadata in translator format
+	 * @return {Promise<boolean>} True if the item was updated and false if not
+	 */
+	combineItemMetadata: async function (item, newItem) {
+		let itemUpdated = false;
+		
+		// Set the new item type
+		let newItemTypeID = Zotero.ItemTypes.getID(newItem.itemType);
+		if (item.itemTypeID !== newItemTypeID) {
+			item.setType(newItemTypeID);
+			itemUpdated = true;
+		}
+		
+		// Check if creators are updated
+		let creators1 = item.toJSON().creators;
+		let creators2 = newItem.creators;
+		
+		let creatorsUpdated = false;
+		
+		// Check if creators count is not equal
+		if (creators1.length !== creators2.length) {
+			creatorsUpdated = true;
+		}
+		// Check if creators aren't equal
+		else {
+			// Compare names and type for each creator
+			for (let i = 0; i < creators1.length; i++) {
+				if (
+					creators1[i].creatorType !== creators2[i].creatorType ||
+					creators1[i].firstName !== creators2[i].firstName ||
+					creators1[i].lastName !== creators2[i].lastName
+				) {
+					creatorsUpdated = true;
+					break;
+				}
+			}
+		}
+		
+		// If at least something is updated in creators, set the new creators
+		if (creatorsUpdated) {
+			// Empty creators, because setCreators() doesn't do that by default
+			item.setCreators([]);
+			item.setCreators(newItem.creators);
+			itemUpdated = true;
+		}
+		
+		// Get all field names for specific item type
+		let fields = Zotero.ItemFields.getItemTypeFields(Zotero.ItemTypes.getID(newItem.itemType));
+		fields = fields.map(x => Zotero.ItemFields.getName(x));
+		
+		// Transfer all fields from the new item to the existing one
+		for (let key of fields) {
+			// Null is when a field doesn't exist
+			let field1 = item.getField(key) || null;
+			let field2 = newItem[key] || null;
+			
+			// If fields aren't identical
+			if (field1 !== field2) {
+				// accessDate is always new, but it doesn't mean that metadata is changed
+				if (key !== 'accessDate') itemUpdated = true;
+				item.setField(key, field2);
```

Comment:
There are many cases when item metadata needs small manual corrections after translation (e.g., correcting capitalization, removing a period from the end of a title). The function as written will overwrite those changes without notice, correct? I worry that that would be very frustrating—should it trigger the resolve conflicts or merge items dialog instead?

Reply:
Along with this metadata updating feature, we plan to also have an undo option to return the previous metadata. But of course it can be a little bit hard to track back what exactly changed, especially if it was a small change like a title fix. It might be convenient to have an option to see all changes for each updated item, but that would be difficult to fit into the current metadata update progress dialog (it looks the same as PDF recognizer dialog). Currently it just shows 'updated' and nothing else.

Reply:
I see that being potentially really frustrating unless it triggers something like the sync conflict dialog to show the changes. 

Updating date, volume, issue, page fields aren’t likely to yield any objections ever. I think the issues might primarily come from title, creators, and book/journal title.

Reply:
I think just showing the revised content in bold would be sufficient. Perhaps double clicking could open something like the merge dialog or the sync conflict dialog?

Reply:
Could you please give an example of how the revised content should be presented in a progress dialog row?

I was thinking that the most convenient way to spot changes would be a diff of two citations that are generated from the old and the updated metadata.

Citation 1:
```
Wise, James B. “Argon Laser Therapy for Open-Angle Glaucoma: A Pilot Study.” Archives of Ophthalmology 97, no. 2 (1979): 319–322.
```

Citation 2:
```
Wise, James B., and Stanton L. Witter. “Argon Laser Therapy for Open-Angle Glaucoma” Archives of Ophthalmology 97, no. 2 (1980): 319–322.
```

Result:
![alt text](https://s3.amazonaws.com/upldat/citations_diff.png)

The citations demonstrate quite common metadata differences: incomplete title, missing authors, different publication date.

But of course it doesn't include changes on other fields that are not typically used in citations i.e. Abstract, URL, etc. But from the convenience perspective that would be the quickest way to review changes, in my opinion.

Merge or conflict dialog can be considered in future if there will be a need.

Reply:
I think that is great. If it uses the default style for Quick Copy or another user-specified style, then it will likely show the the most relevant differences for most users.

Reply:
Yeah, I think that's really nice. I assume this would need to go in a space below each row?

It would take some work, but ideal might be an edit button that opens a panel (like what's used for Add Item by Identifier) with the merge-items interface, that let you choose individual versions of fields. I can't remember how generalizable that is right now — if it takes too much work, it'd be better to reimplement it with React (which we're hopefully almost able to do).

Probably should also be possible to revert with a single button.